### PR TITLE
Insert 3.1.6 runtime

### PR DESCRIPTION
**Pull request by wtgodbe**
This will need a stage0 from the 3.1.6 insertion into dotnet/installer. I think installer will need to ingest CLI & Toolset with the updated 3.1.6 runtime as well before the produced SDK will unblock this PR

### CODE REVIEW

**Review by joeloff — APPROVED** [review]

**Comment by joeloff** [comment]
yeah, we'll need stage 0 in the CLI too I think

**Comment by wtgodbe** [comment]
https://dev.azure.com/dnceng/internal/_build/results?buildId=731548&view=results should produce the stage0

**Comment by wtgodbe** [comment]
https://github.com/dotnet/installer/pull/8010 will produce the stage0

**Comment by wtgodbe** [comment]
Stage0 pushed

**Comment by wtgodbe** [comment]
@joeloff @marcpopMSFT this is failing because the stage0 SDK has implicit version 2.1.21, which isn't available yet. Should we merge with failing tests (since the product will still build fine), or back the implicit version change out of installer, get a new stage0, then re-do the implict version change?

**Comment by joeloff** [comment]
What is the implicit reference, or rather where is it coming from?

**Comment by wtgodbe** [comment]
It comes from here: https://github.com/dotnet/installer/blob/1429d6653e297e35cdf56262404769107368d219/src/redist/targets/GenerateBundledVersions.targets#L92-L107

dotnet/installer's 3.1 branches reference both the latest 3.1 & 2.1 templates, but since 3.1 & 2.1 release on the same day, 3.1 will reference a 2.1 version that isn't published yet up until release day. We get around this in dotnet/installer by turning off some tests - we could do the same here, if we think Stage0's will always also contain implicit version updates (we recently changed our behavior to do the implicit version update earlier in the timeline). The other option is to stop doing implicit version updates concurrently with branding.

**Comment by joeloff** [comment]
But we'd still include it, it's just that we can't download the packages during the build, because we should be including them when inserting. So with your suggestion we'd be referencing the latest available version (as opposed to the eventual latest version)?

**Comment by wtgodbe** [comment]
We need to reference the 'eventual' version in dotnet/installer - it's just a matter of whether we should update that reference _after_ we create stage0s (which may be hard to predict), or we allow tests related to the 2.1 implicit version to fail in this repo when using a stage0. I think the latter is probably fine

**Comment by joeloff** [comment]
I think I'm okay with what you're suggesting. We'd just need to keep an eye out for the test failures. I assume the failures wouldn't block publishing anything? Also, by the time we're inserting this into 16.7, we'd need to take a 2.1.21 insertion as well I assume.

**Comment by wtgodbe** [comment]
That's correct, the test failures won't block publishing. The issue actually isn't in this repo, it's in the SDK we produced out of dotnet/installer, which will continue to be unusable until we publish 2.1.21. That's just an unfortunate consequence of needing to update the implicit versions ahead of time. I'll go ahead and merge this.

**Comment by wtgodbe** [comment]
Ah, I forgot this repo blocks merging without a green CI. So we'll need to either disable that restriction, or back out the implicit version update from installer, for the purpose of getting a new stage0 (of course we'd redo the implicit version update after that). @marcpopMSFT @wli3 @sfoslund @dsplaisted thoughts?

**Comment by wli3** [comment]
> back out the implicit version update from installer

It should not block getting a stage 0. If so, we can disable the test blocking it.

**Comment by wtgodbe** [comment]
There are 10-15 failing test cases - should I disable all of them?

**Comment by joeloff** [comment]
They seem to be all 2.1 related, so that's probably the right thing to do

**Comment by wtgodbe** [comment]
It was quite a few tests: https://github.com/dotnet/sdk/pull/12460/commits/16561beb133e2438fa4e05b16dcd376a0d5d33cc

**Review by joeloff — APPROVED** [review]

**Comment by joeloff** [comment]
I'm wondering whether we should back this out of 3.1.4xx. We need to ingest the 3.1.6 runtime for 3.1.400, so the matching 2.1 bits should be 2.1.20, not 2.1.21 because that becomes available the same day when we already have to be in VS 16.7 GA

**Comment by wtgodbe** [comment]
> I'm wondering whether we should back this out of 3.1.4xx. We need to ingest the 3.1.6 runtime for 3.1.400, so the matching 2.1 bits should be 2.1.20, not 2.1.21 because that becomes available the same day when we already have to be in VS 16.7 GA

Oh, very good point. 2.1.21 won't be available when we publish 3.1.400. Backing out the implicit version update seems right - @joeloff can you create such a PR?

**Comment by joeloff** [comment]
yes, I'll prep a PR

**Comment by wtgodbe** [comment]
Correct stage0: https://dev.azure.com/dnceng/internal/_build/results?buildId=740809&view=results

File: global.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "tools": {
-    "dotnet": "3.1.400-preview-015204",
+    "dotnet": "3.1.400-preview-015212",
```
Comment (joeloff) [inline global.json:3]: Look at the build link you shared, shouldn't this be 015217?

Comment (wtgodbe) [reply global.json:3]: Yeah, once the linked build finishes. I'll update this PR once it's done (and I'll back out the test changes)

**Comment by wtgodbe** [comment]
Undid the test changes

**Pull request merged**